### PR TITLE
Optimization to delay creation of Hollow Objects only when needed.

### PR DESCRIPTION
**Pull request by davesu**
- Enhance computeDataChange to capture ordinals only and not create Hollow Objects
- Benefits will be significant for large state changes

### CODE REVIEW

File: hollow/src/main/java/com/netflix/hollow/api/consumer/data/AbstractHollowDataAccessor.java [hunk 1]
```diff
@@ -99,59 +102,33 @@ protected void computeDataChange(String type, HollowReadStateEngine stateEngine,
...
 
                 addedOrdinal = addedOrdinals.nextSetBit(addedOrdinal + 1);
             }
         }
-
-        { // determine removed records
-            int removedOrdinal = removedOrdinals.nextSetBit(0);
-            while (removedOrdinal != -1) {
-                // exclude records that was removed but re-added with different ordinal since that is considered as updated record
-                if (!updatedRecordOrdinals.get(removedOrdinal)) {
```
Comment (Sunjeet) [inline hollow/src/main/java/com/netflix/hollow/api/consumer/data/AbstractHollowDataAccessor.java:143]: This logic might be needed to correctly compute removed ordinals, without it updates will also be returned as removed

**Comment by Sunjeet** [comment]
Is this cost currently incurred in delta application? Does moving the deserialization cost to access time move the problem elsewhere? Do getAllRecords() etc. get called repeatedly, are the deserialized records cached?

File: hollow/src/main/java/com/netflix/hollow/api/consumer/data/AbstractHollowDataAccessor.java (hunk 1 above)
Comment (davesu) [reply hollow/src/main/java/com/netflix/hollow/api/consumer/data/AbstractHollowDataAccessor.java:143]: yeah, I did notice the issue with the unit test but was not able to run it locally due to build/dependency (jcenter repo) issues.  

Now that the above jcenter repo issue was corrected, I was able to run unit test and fix the issues.  Please take another look at the PR.

**Comment by davesu** [comment]
> Is this cost currently incurred in delta application? Does moving the deserialization cost to access time move the problem elsewhere? Do getAllRecords() etc. get called repeatedly, are the deserialized records cached?

Most people don't currently use the generated DataAccessor.  Also one will not notice performance overhead with small dataset or small deltas.  But with large datasets creating all those POJO during change calculate could be really costly both CPU and large heap.   Specially for use cases which only cares about modified records and not added nor removed ones.   

Therefore, this change falls inline with hollow practice on only hydrating POJOs when needed.

File: hollow/src/main/java/com/netflix/hollow/api/consumer/data/AbstractHollowDataAccessor.java [hunk 2]
```diff
@@ -27,19 +27,22 @@
...
+import java.util.Objects;
 
 public abstract class AbstractHollowDataAccessor<T> {
     protected final String type;
     protected final PrimaryKey primaryKey;
     protected final HollowReadStateEngine rStateEngine;
 
-    private List<T> removedRecords = Collections.emptyList();
-    private List<T> addedRecords = Collections.emptyList();
+    private BitSet removedOrdinals = new BitSet();
```
Comment (Sunjeet) [inline hollow/src/main/java/com/netflix/hollow/api/consumer/data/AbstractHollowDataAccessor.java:44]: nit: initialized twice with `= new BitSet()`, we can drop it from one place

**Review by Sunjeet — APPROVED** [review]

**Pull request merged**